src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractAddItemCommandHandler.cs

```diff
@@ -0,0 +1,131 @@
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Input.Commands
+{
+    /// <summary>
+    /// Provides support for all Add Item commands that operate on <see cref="IProjectTree"/> nodes, across C# and VB
+    /// </summary>
+    internal abstract partial class AbstractAddItemCommandHandler : IAsyncCommandGroupHandler
+    {
+        protected static readonly Guid LegacyCSharpPackageGuid = new Guid("{FAE04EC1-301F-11d3-BF4B-00C04F79EFBC}");
+        protected static readonly Guid LegacyVBPackageGuid = new Guid("{164B10B9-B200-11d0-8C61-00A0C91E29D5}");
+
+        private readonly IPhysicalProjectTree _projectTree;
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private readonly IVsUIService<IVsAddProjectItemDlg> _addItemDialog;
+        private readonly IVsService<SVsShell, IVsShell> _vsShell;
+        private readonly Lazy<Dictionary<long, CommandDetails>> _cSharpCommandMap;
+        private readonly Lazy<Dictionary<long, CommandDetails>> _vbCommandMap;
+
+        [ImportingConstructor]
+        public AbstractAddItemCommandHandler(IPhysicalProjectTree projectTree, IUnconfiguredProjectVsServices projectVsServices, IVsUIService<IVsAddProjectItemDlg> addItemDialog, IVsService<SVsShell, IVsShell> vsShell)
+        {
+            _projectTree = projectTree;
+            _projectVsServices = projectVsServices;
+            _addItemDialog = addItemDialog;
+            _vsShell = vsShell;
+
+            _cSharpCommandMap = new Lazy<Dictionary<long, CommandDetails>>(() => GetCSharpCommands());
+            _vbCommandMap = new Lazy<Dictionary<long, CommandDetails>>(() => GetVBCommands());
+        }
+
+        protected abstract Dictionary<long, CommandDetails> GetCSharpCommands();
+        protected abstract Dictionary<long, CommandDetails> GetVBCommands();
+
+        public Task<CommandStatusResult> GetCommandStatusAsync(IImmutableSet<IProjectTree> nodes, long commandId, bool focused, string commandText, CommandStatus progressiveStatus)
+        {
+            Requires.NotNull(nodes, nameof(nodes));
+
+            if (TryGetCommandDetails(commandId, out _))
+            {
+                return GetCommandStatusResult.Handled(commandText, CommandStatus.Enabled);
+            }
+
+            return GetCommandStatusResult.Unhandled;
+        }
+
+        public async Task<bool> TryHandleCommandAsync(IImmutableSet<IProjectTree> nodes, long commandId, bool focused, long commandExecuteOptions, IntPtr variantArgIn, IntPtr variantArgOut)
+        {
+            Requires.NotNull(nodes, nameof(nodes));
+
+            if (nodes.Count == 1)
+            {
+                if (TryGetCommandDetails(commandId, out CommandDetails result))
+                {
+                    IProjectTree node = nodes.First();
+
+                    // Get the strings from the legacy package
+                    IVsShell vsShell = await _vsShell.GetValueAsync();
```

Comment:
While commands are likely to be called on the UI thread, you should also explicitly switch to the UI thread anyway before touching the shell.